File: solution/874. Walking Robot Simulation/874.cpp
class Solution {
public:
    int robotSim(vector<int>& commands, vector<vector<int>>& obstacles) {
        set<pair<int, int>> O;
        for (auto& o : obstacles)
            O.insert({o[0], o[1]});
        const vector<pair<int, int>> dirs = {{0, 1}, {1, 0}, {0, -1}, {-1, 0}};
        
        int res = 0;
        int y = 0, x = 0, d = 0;
        for (auto& k : commands) {
            if (k > 0) {
                while (k--) {
                    if (O.count({x + dirs[d].first, y + dirs[d].second})) break;
                    x += dirs[d].first;
                    y += dirs[d].second;
                }
            } else if (k == -2) {
                d = (d - 1 + 4) % 4;
            } else if (k == -1) {
                d = (d + 1) % 4;
            }
            res = max(res, x * x + y * y);
        }
        return res;
    }
};

File: solution/874. Walking Robot Simulation/874.py
class Solution:
    def robotSim(self, commands: List[int], obstacles: List[List[int]]) -> int:
        dirs = [(0, 1), (1, 0), (0, -1), (-1, 0)]
        obstacles = set(map(tuple, obstacles))
        res = y = x = d = 0
        for c in commands:
            if c == -1:
                d = (d + 1) % 4
            elif c == -2:
                d = (d - 1) % 4
            else:
                dx, dy = dirs[d]
                while c and (x + dx, y + dy) not in obstacles:
                    x += dx
                    y += dy
                    c -= 1
            res = max(res, x * x + y * y)
        return res
    
